### PR TITLE
feat(network): Add `eth` handshake metrics

### PR DESCRIPTION
**Pull request by emnul**
Adds `eth` handshake metrics described in https://github.com/paradigmxyz/reth/issues/3665 

Ref #3665

### CODE REVIEW

**Review by onbjerg — DISMISSED** [review]
please remove the redundant semicolons:)

File: crates/net/network/src/metrics.rs [hunk 1]
```diff
@@ -81,6 +81,32 @@ pub struct NetworkMetrics {
 #[derive(Metrics)]
 #[metrics(scope = "network")]
 pub struct SessionManagerMetrics {
+    /// Number of errors related to incoming incorrect peer behaviour, such as invalid message
+    /// code, size, encoding, etc.
+    pub(crate) total_incoming_eth_handshake_peer_errors: Counter,
+    /// Number of errors related to outgoing incorrect peer behaviour, such as invalid message
+    /// code, size, encoding, etc.
+    pub(crate) total_outgoing_eth_handshake_peer_errors: Counter,
+    /// Number of incoming `StreamTimeout` errors
```
[the remark this answers is not in the document]
Comment (emnul) [reply crates/net/network/src/metrics.rs]: Thanks for this extra info! Thought these were dev facing. I've updated them accordingly.

**Comment by emnul** [comment]
@onbjerg I don't think this PR should close https://github.com/paradigmxyz/reth/issues/3665 because it's just one part of many that still need to be added along with changes to the Grafana dashboard to display the new metrics. I plan on tackling the other metrics in separate PRs.

File: crates/net/network/src/metrics.rs [hunk 2]
```diff
@@ -81,6 +81,32 @@ pub struct NetworkMetrics {
...
+    /// Number of outgoing protocol version mismatch errors
+    pub(crate) total_outgoing_eth_handshake_version_errors: Counter,
+    /// Number of incoming genesis block mismatch errors
+    pub(crate) total_incoming_eth_handshake_genesis_errors: Counter,
+    /// Number of outgoing genesis block mismatch errors
+    pub(crate) total_outgoing_eth_handshake_genesis_errors: Counter,
+    /// Number of incoming fork id mismatch errors
+    pub(crate) total_incoming_eth_handshake_forkid_errors: Counter,
+    /// Number of outgoing fork id mismatch errors
+    pub(crate) total_outgoing_eth_handshake_forkid_errors: Counter,
```
Comment (emhane) [inline crates/net/network/src/metrics.rs:109]: names need to take into account https://github.com/paradigmxyz/reth/issues/8150

**Review by emhane — CHANGES_REQUESTED** [review]

**Comment by emnul** [comment]
@onbjerg After updating the Peer Disconnect Reasons panel, I copy/pasted the updated JSON Model from Dashboards -> Reth Transaction Pool -> Settings -> JSON Model in the Grafana UI into `etc/grafana/dashboards/reth-mempool.json` not sure if this is the recommended way to update these files. 

Also I don't know how I might simulate Peer Disconnect errors. Right now, I'm running a private network with Kurtosis with the following `network_params.yml`
```yaml
participants:
   - el_type: reth
     el_image: reth_v1.0.6_update-reth-mempool-dashboard
     cl_type: lighthouse
     cl_image: sigp/lighthouse:latest
   - el_type: reth
     el_image: reth_v1.0.6_update-reth-mempool-dashboard
     cl_type: teku
     cl_image: consensys/teku:latest
additional_services:
   - prometheus_grafana
   - apache
grafana_additional_dashboards:
   [
      "path/to/etc/grafana/dashboards/reth-mempool.json",
      "path/to/etc/grafana/dashboards/overview.json",
   ]

```
But I'm not seeing any disconnect errors (which I think makes sense). If you or any of the other maintainers know a better way to test this please let me know. Definitely want to be as thorough as possible with my testing.

Screenshot 2024-08-30 at 17 26 37

**Review by emhane — APPROVED** [review]

**Comment by emnul** [comment]
The changes to `etc/grafana/dashboards/reth-mempool.json` result in the following when running a reth node via the cli. I'm guessing the exported JSON from my local Grafana setup with Kurtosis, included some Kurtosis specific configs that don't work out of the box when running Reth standalone. 

Screenshot 2024-08-30 at 18 42 11

**Comment by emnul** [comment]
Ok the latest commit yields this dashboard running a reth node via the cli. I think this is what we want

Screenshot 2024-08-30 at 19 10 28

**Comment by emnul** [comment]
Still looking for some guidance on the best approach to update the dashboard .json file to ensure it works for all users rather than my specific configuration.

**Comment by emnul** [comment]
Totally missed that, thanks @emhane !

File: etc/grafana/dashboards/reth-mempool.json [hunk 3]
```diff
@@ -1,3529 +1,3666 @@
...
-      },
-      "gridPos": {
-        "h": 3,
-        "w": 6,
-        "x": 3,
-        "y": 1
+      {
+         "datasource": {
+            "type": "prometheus",
+            "uid": "edwfgevdigk5cf"
```
Comment (emhane) [inline etc/grafana/dashboards/reth-mempool.json:44]: when you have copied the json provided from sharing for external use, then the uid should be `"${DS_PROMETHEUS}"`

**Review by emhane — CHANGES_REQUESTED** [review]

**Review by onbjerg — CHANGES_REQUESTED** [review]
ping @emnul, please re-export the dashboard but with the "share for external use" checkbox checked so it replaces the prometheus references

**Comment by emnul** [comment]
Sorry for the delay! I've updated the dashboard using the "Export for Sharing Externally" method

**Review by emhane — CHANGES_REQUESTED** [review]
the transaction pool row shouldn't be collapsed, otherwise looks like no panels on main missing (a common mistake, grafana is a bit tricky to update)

Screenshot 2024-09-11 at 11 42 18

**Comment by emnul** [comment]
Wow didn't even realize that was configurable haha making a quick update now.

File: crates/net/network/src/metrics.rs
Comment (mattsse) [inline crates/net/network/src/metrics.rs:1]: are all of these really meaningful?

File: crates/net/network/src/session/mod.rs [hunk 4]
```diff
@@ -552,12 +554,72 @@ impl SessionManager {
                 self.remove_pending_session(&session_id);
                 match direction {
                     Direction::Incoming => {
+                        match error.as_ref().and_then(|e| e.as_eth_stream()) {
```
Comment (mattsse) [inline crates/net/network/src/session/mod.rs:557]: this is just a match over an error, right?

so I'd like to extract this to a function instead

File: crates/net/network/src/session/mod.rs [hunk 5]
```diff
@@ -552,12 +554,72 @@ impl SessionManager {
...
+                            _ => self.metrics.incoming_eth_handshake_peer_error_total.increment(1),
+                        }
                         Poll::Ready(SessionEvent::IncomingPendingSessionClosed {
                             remote_addr,
                             error,
                         })
                     }
                     Direction::Outgoing(peer_id) => {
+                        match error.as_ref().and_then(|e| e.as_eth_stream()) {
+                            Some(EthStreamError::StreamTimeout) => {
```
Comment (mattsse) [inline crates/net/network/src/session/mod.rs:594]: same

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by emhane** [comment]
I can't see any new panels added by this pr? the tx pool row is still not dropped down when I build this dashboard

**Comment by mattsse** [comment]
what's the path forward here @emnul @emhane

File: crates/net/network/src/session/mod.rs (hunk 4 above)
Comment (emhane) [reply crates/net/network/src/session/mod.rs:557]: agreed, smthg like `SessionManager::update_disconnect_incoming_metrics`

File: etc/grafana/dashboards/reth-mempool.json
Comment (emhane) [inline etc/grafana/dashboards/reth-mempool.json:1]: let's just check out this file from main, and add panels in a separate pr

**Review by emhane — CHANGES_REQUESTED** [review]

**Comment by emnul** [comment]
@emhane

>  I can't see any new panels added by this pr?
I had added the metrics to the "Peer Disconnect Reasons" panel. Do we want them in their own panel / pie graph?

> the tx pool row is still not dropped down when I build this dashboard
Sorry about that, I thought I had fixed it 🤔  I'll take another look!

> what's the path forward here 
@mattsse sorry I had a bit of a setback last week when I had to wipe my whole OS and didn't have any backups of my machine rip. I should be able to make the recommended changes by Thursday.

**Comment by emhane** [comment]
> I had added the metrics to the "Peer Disconnect Reasons" panel. Do we want them in their own panel / pie graph?
> 
I'll check again, seems I missed them, apologies. would you still mind leaving it out of this pr, and updating that panel in a separate pr?

**Review by mattsse — COMMENTED** [review]
imo none of those are useful, the existing connection metrics are sufficient

**Pull request closed without merging**